File: python/day3.py
from utilities import read_lines
from functools import reduce


def count_trees(data, moves):
    trees = 0
    row = 0
    col = 0
    while row < len(data):
        if data[row][col] == '#':
            trees += 1
        row += moves[1]
        col += moves[0]
        col %= len(data[0])
    return trees


def part_1(data):
    return count_trees(data, [3, 1])


def part_2(data):
    moves = [[1, 1], [3, 1], [5, 1], [7, 1], [1, 2]]
    return reduce(lambda x, y: x * y, [count_trees(data, move) for move in moves])


if __name__ == '__main__':
    print('Day 3')
    data = read_lines('../inputs/day3.txt')
    print('\tPart 1: {}'.format(part_1(data)))
    print('\tPart 2: {}'.format(part_2(data)))

File: rust/src/day3.rs
use crate::utilities::read_lines;

fn count_trees(data: &Vec<Vec<char>>, move_: &(usize, usize)) -> i32 {
    let mut trees = 0;
    let mut row = 0;
    let mut col = 0;

    while row < data.len() {
        if data[row][col] == '#' {
            trees += 1;
        }
        row += move_.1;
        col += move_.0;
        col %= data[0].len();
    }

    trees
}

fn part_1(data: &Vec<Vec<char>>) {
    let move_ = (3, 1);
    println!("\tPart 1: {}", count_trees(data, &move_));
}

fn part_2(data: &Vec<Vec<char>>) {
    let moves = vec![(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)];
    println!(
        "\tPart 2: {}",
        moves
            .iter()
            .map(|move_| count_trees(data, move_))
            .fold(1, |x, y| x * y)
    );
}

pub fn main() {
    println!("Day 3");
    let data: Vec<Vec<char>> = read_lines("../inputs/day3.txt")
        .into_iter()
        .map(|line| line.chars().collect())
        .collect();
    part_1(&data);
    part_2(&data);
}
